cpp/src/arrow/array/builder_nested.cc

```diff
@@ -62,31 +66,23 @@ Status MapBuilder::Resize(int64_t capacity) {
 
 void MapBuilder::Reset() {
   list_builder_->Reset();
-  item_builder_->Reset();
   ArrayBuilder::Reset();
 }
 
 Status MapBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
   ARROW_CHECK_EQ(item_builder_->length(), key_builder_->length())
       << "keys and items builders don't have the same size in MapBuilder";
-  // finish list(keys) builder
+  RETURN_NOT_OK(CheckStructBuilderLength());
   RETURN_NOT_OK(list_builder_->FinishInternal(out));
-  // finish values builder
-  std::shared_ptr<ArrayData> items_data;
-  RETURN_NOT_OK(item_builder_->FinishInternal(&items_data));
-
-  auto keys_data = (*out)->child_data[0];
   (*out)->type = type();
-  (*out)->child_data[0] =
-      ArrayData::Make((*out)->type->child(0)->type(), keys_data->length, {nullptr},
-                      {keys_data, items_data}, 0, 0);
```

Comment:
This seemed a bit odd, using a list of keys builder and a separate item builder. I changed it to use use a list builder with a child struct builder. This way I could use this builder to convert a py sequence with the ListConverter as-is.  The only issue was that when using the separate key_builder and item_builders, the struct_builder needs a call to `Append`\`AppendValues` before adding the next index in the MapArray.